libcst/metadata/scope_provider.py

```diff
@@ -0,0 +1,356 @@
+        return self._visit_import_alike(node)
+
+    def visit_Name(self, node: cst.Name) -> Optional[bool]:
+        context = self.provider.get_metadata(ExpressionContextProvider, node)
+        if context == ExpressionContext.STORE:
+            self.scope.record_assignment(node.value, node)
+        elif context == ExpressionContext.LOAD:
+            self.__deferred_accesses.append(Access(node, self.scope))
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> Optional[bool]:
+        self.scope.record_assignment(node.name.value, node)
+        with self._new_scope(FunctionScope):
+            node.params.visit(self)
+            node.body.visit(self)
+
+        for decorator in node.decorators:
+            decorator.visit(self)
+        if node.returns:
+            node.returns.visit(self)
+        return False
+
+    def visit_Lambda(self, node: cst.Lambda) -> Optional[bool]:
+        with self._new_scope(FunctionScope):
+            node.params.visit(self)
+            node.body.visit(self)
+        return False
+
+    def visit_Param(self, node: cst.Param) -> Optional[bool]:
+        self.scope.record_assignment(node.name.value, node)
+        with self._switch_scope(self.scope.parent):
+            for field in [node.default, node.annotation]:
+                if field:
+                    field.visit(self)
+        return False
+
+    def visit_Arg_keyword(self, node: cst.Arg) -> None:
+        if node.keyword is not None:
+            self.scope.record_assignment(node.keyword.value, node)
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> Optional[bool]:
+        self.scope.record_assignment(node.name.value, node)
+        for decorator in node.decorators:
+            decorator.visit(self)
+        for base in node.bases:
+            base.visit(self)
+        for keyword in node.keywords:
+            keyword.visit(self)
+
+        with self._new_scope(ClassScope):
+            for statement in node.body.body:
+                statement.visit(self)
+        return False
+
+    def visit_Global(self, node: cst.Global) -> Optional[bool]:
+        for name_item in node.names:
+            self.scope.record_global_overwrite(name_item.name.value)
+        return False
+
+    def visit_Nonlocal(self, node: cst.Nonlocal) -> Optional[bool]:
+        for name_item in node.names:
+            self.scope.record_nonlocal_overwrite(name_item.name.value)
+        return False
+
+    def visit_ListComp(self, node: cst.ListComp) -> Optional[bool]:
+        return self._visit_comp_alike(node)
+
+    def visit_SetComp(self, node: cst.SetComp) -> Optional[bool]:
+        return self._visit_comp_alike(node)
+
+    def visit_DictComp(self, node: cst.DictComp) -> Optional[bool]:
+        return self._visit_comp_alike(node)
+
+    def visit_GeneratorExp(self, node: cst.GeneratorExp) -> Optional[bool]:
+        return self._visit_comp_alike(node)
+
+    def _visit_comp_alike(
```

Comment:
It's clear what this code does, but it's not clear why it does it, and IMO, there's a lot of potential landmines in comprehension scopes.

Can you copy some of the comments in the AST implementation here?

Reply:
@bgw 
Added docstring for more explanation.

In the original comment regarding nested comprehension, it mentioned the scope of outer iterator node (CompFor) will be the ComprehensionScope while the node.iter will be parent scope. It looked not straightforward to me. I thought it's more intuitive if the child node always extends the parent scope or create a new scope rather than extend grandfather scope. So I made the decision to set the CompFor node to parent scope.